### ISSUE TITLE
[ARCHETYPE-492] Underscore in filenames problematic due to greedy regex

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-archetype#763]
**[Oliver Dungey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oliver.dungey)** opened **[ARCHETYPE-492](https://issues.apache.org/jira/browse/ARCHETYPE-492?redirect=false)** and commented

If you have a file in a project with an underscore adjacent to a substitution variable, the substitution will fail i.e.

if `artifactId` is 'Test' and a target file is to be named `something_test.txt` the file in the template should be named `something____artifactId__.txt` (3 underscores in the middle) but the result is an error stating that the property `_artifactId` cannot be found. This is because the term inside the regex is a greedy .*  - a simple fix would be to change this to something like __[^_]*___  which would only match non-underscore characters.

Fixing this issue would allow the use of underscores in filenames in all circumstances rather than the current situation where you may get lucky. The only down side to fixing this issue is that properties with leading or trailing underscores will not be valid - this seems a far more preferable situation.

A patch for this issue was put on this bug back in 01//2009 but somehow got ignored - see [Alexander Maslov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maslovalex) patch and comments at the end of this issue: https://issues.apache.org/jira/browse/ARCHETYPE-191

Apologies for the formatting - it appears double underscores are an issue :-)

---

**Affects:** 2.4

**Issue Links:**
- [ARCHETYPE-400](https://issues.apache.org/jira/browse/ARCHETYPE-400) method DefaultFilesetArchetypeGenerator.replaceFilenameTokens(final String filePath, final Context context) doesn't work when two tokens are close to each other
 (_**"fixes"**_)

5 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Matthias Hanisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=matze)** commented

Got the same problem today. Wanted to generate a file

```
errors_${artifactId}.properties
```

so I chose the filename

```
errors___artifactId__.properties
```

in the archetype.
The placeholder in the file name was not resolved.

**Comment by jira-importer** [comment]
**[Matthias Hanisch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=matze)** commented

Just to let everybody else know my workaround:

In archetype-metadata I created a requiredProperty like this:

```
<requiredProperty key="underscoreArtifactId">
        <defaultValue>_${artifactId}</defaultValue>
</requiredProperty>
```

The above filename can be now written as

```
errors__underscoreArtifactId__.properties
```

This avoids the triple-underscore-problem. Ugly, but works.

**Comment by jira-importer** [comment]
**[Giovanni Lovato](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=heruan)** commented

This is an issue also for archetypes which contains Flyway database migration files, which has a specific filename pattern, e.g. `V2\_\_my-schema.sql` (i.e. version + `\_` + `\_` + description). So if I include in the archetype a file named `V2\_\_\_\_rootArtifactId\_\_-schema.sql` I would expect it to be created as `V2\_\_my-artifact-schema.sql`.

**Comment by jira-importer** [comment]
**[Guillaume Boué](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gboue)** commented

FIxed in [bb723acb42efcffd335575499ac2569f98dbbaa9](http://git-wip-us.apache.org/repos/asf/maven-archetype/commit/bb723acb42efcffd335575499ac2569f98dbbaa9).
